solver/src/solver/paraswap_solver/api.rs

```diff
@@ -29,18 +29,36 @@ pub struct DefaultParaswapApi {
 
 #[async_trait::async_trait]
 impl ParaswapApi for DefaultParaswapApi {
-    async fn price(&self, query: PriceQuery) -> Result<PriceResponse> {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse, ParaswapResponseError> {
+        let query_str = format!("{:?}", &query);
         let url = query.into_url();
         tracing::debug!("Querying Paraswap API (price) for url {}", url);
-        let text = reqwest::get(url)
-            .await
-            .context("PriceQuery failed")?
-            .text()
-            .await?;
-        tracing::debug!("Response from Paraswap API (price): {}", text);
-
-        serde_json::from_str::<PriceResponse>(&text)
-            .context(format!("PriceQuery result parsing failed: {}", text))
+        let response_result = reqwest::get(url).await;
+        match response_result {
+            Ok(response) => match response.text().await {
+                Ok(response_text) => {
+                    tracing::debug!("Response from Paraswap API (price): {}", response_text);
+                    let raw_response = serde_json::from_str::<RawPriceResponse>(&response_text);
+                    match raw_response {
+                        Ok(RawPriceResponse::ResponseOk(response)) => Ok(response),
+                        Ok(RawPriceResponse::ResponseErr { error: message }) => {
+                            match &message[..] {
+                                "computePrice Error" => Err(ParaswapResponseError::ComputePrice(
+                                    query_str.parse().unwrap(),
+                                )),
+                                _ => Err(ParaswapResponseError::UnknownParaswapError(format!(
+                                    "uncatalogued Price Query error message {}",
+                                    message
+                                ))),
+                            }
+                        }
+                        Err(err) => Err(ParaswapResponseError::DeserializeError(err)),
+                    }
+                }
+                Err(text_fetch_err) => Err(ParaswapResponseError::TextFetch(text_fetch_err)),
+            },
+            Err(send_err) => Err(ParaswapResponseError::Send(send_err)),
```

Comment:
I feel like there is quite a bit of right-ward drift in this block. I would prefer to use something like `map_err` so that you can use the `?` try operator:
```rust
reqwest::get(url)
    .await
    .map_err(ParaswapResponseError::Send)?
    .text()
    .await
    .map_err(ParaswapResponseError::TextFetch)?
    // ...
```

Reply:
Yes! I was hoping someone would come around and make this kinda suggestion!

Reply:
Fixed this, thanks so much!